[FIX] l10n_latam_invoice_document: Allow to change the l10n_latam_document…

suv-bmya opened the pull request:
…_number manually when the invoice is in draft status

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:




--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/70718).

suv-bmya commented:
@jco-odoo here the first approach to allow change the l10n_latam_document_number manually. 

I removed the condition related with the l10n_latam_manual_document_number and highest_name from the model and it works well. On the other hand, I've removed the same fields from the view but I had to remove the l10n_latam_manual_document_number too because l10n_latam_manual_document_number is set to False once the first invoice is created.
Finally,  I also removed the posted_before field to readonly attribute but I'm not sure if we want to allow change the document number once the invoice is posted.

Take a look, please. I would like discuss it tomorrow in the meeting.

jco-odoo commented on addons/l10n_latam_invoice_document/views/account_move_view.xml:
```diff
@@ -45,7 +45,7 @@
                     domain="[('id', 'in', l10n_latam_available_document_type_ids)]" options="{'no_open': True, 'no_create': True}"/>
                 <field name="l10n_latam_document_number"
-                    attrs="{'invisible': ['|', ('l10n_latam_use_documents', '=', False), ('l10n_latam_manual_document_number', '=', False), '|', '|', ('l10n_latam_use_documents', '=', False), ('highest_name', '!=', False), ('state', '!=', 'draft')], 'required': ['|', ('l10n_latam_manual_document_number', '=', True), ('highest_name', '=', False)], 'readonly': [('posted_before', '=', True), ('state ...
+                    attrs="{'invisible': [('l10n_latam_use_documents', '=', False)], 'required': ['|', ('l10n_latam_manual_document_number', '=', True), ('highest_name', '=', False)]}"/>
```
@suv-bmya It does need to stay read-only in some cases.  This is quite a rough change

vbe-odoo commented:
@vbe-odoo do FUN test on this one

vbe-odoo commented:
Hi @suv-bmya @Danisan I am a bit lost on this one, maybe you can help me out.
I am testing it and I can only change the l10n_latam_document_number on the Vendor Bill side, but not on the Invoice Side.
Should this happen on both sides?
In both cases, I have created an Invoice/Vendor Bill first, and then on the 2nd, 3rd document created tried to see if the l10n_latam_document_number was visible and editable. And this only happened on the Vendor Bill side.

Thanks!
best
@ren-odoo

suv-bmya commented:
> Hi @suv-bmya @Danisan I am a bit lost on this one, maybe you can help me out.
> I am testing it and I can only change the l10n_latam_document_number on the Vendor Bill side, but not on the Invoice Side.
> Should this happen on both sides?
> In both cases, I have created an Invoice/Vendor Bill first, and then on the 2nd, 3rd document created tried to see if the l10n_latam_document_number was visible and editable. And this only happened on the Vendor Bill side.
> 
> Thanks!
> best
> @ren-odoo

@vbe-odoo , What runbot did you use to test it?

vbe-odoo commented:
Hi @suv-bmya I've used the one attahed to this PR. Details below:
Runbot database: https://8899933-14-0-all.runbot58.odoo.com/
Runbot branch & Builds: https://runbot.odoo.com/runbot/r-d-1?search=bmya%3A14.0_allow_change_l10n_latam_document_number

Thanks!

suv-bmya commented:
> Hi @suv-bmya I've used the one attached to this PR. Details below:
> Runbot database: https://8899933-14-0-all.runbot58.odoo.com/
> Runbot branch & Builds: https://runbot.odoo.com/runbot/r-d-1?search=bmya%3A14.0_allow_change_l10n_latam_document_number
> 
> Thanks!

Thanks @vbe-odoo . 

To change the document number you should confirm the invoice. The flow is:
* Create draft invoice
* Confirm invoice: the document number is assigned automatically
* Set the invoice to draft: the document number is visible and you can edit it

![image](https://user-images.githubusercontent.com/52895221/126206236-8ef1a871-3a10-4ed9-bc89-e6f6c6fceb2f.png)

vbe-odoo commented:
Hey, @suv-bmya Oh I see! I was missing that step, the reset to draft.
I tested for the CL company and it's working OK.
I also installed AR just in case and tested it, and didn't affect how AR is working, so everything seems OK.

Just to keep you in the loop @jjscarafia this PR is to be able to change the sequence in customer invoices in CL.

This is ready for technical review @jco-odoo 
Thanks!

jjscarafia commented:
Thanks @vbe-odoo !

jco-odoo commented on addons/l10n_cl/models/account_move.py:
```diff
@@ -124,3 +124,19 @@ def _get_name_invoice_report(self):
         return super()._get_name_invoice_report()
+
+    @api.onchange('l10n_latam_document_type_id', 'l10n_latam_document_number')
+    def _inverse_l10n_latam_document_number(self):
```
After a rebase it should work without overriding this method. @suv-bmya

suv-bmya replied:
👍

jco-odoo commented on addons/l10n_cl/views/account_move_view.xml:
```diff
@@ -15,6 +15,31 @@
+                    	               ('state', '!=', 'draft'),
+                    	               ('country_code', '!=', 'CL')
+                    ],
+                    'readonly': [('posted_before', '=', True), ('state', '!=', 'draft'), ('country_code', '!=', 'CL')]
```
As the readonly is True only when the state is not draft, I don't know why we add the extra condition country_code != 'CL'

Also, why do we remove the "required" one or is it still kept magically?

suv-bmya replied:
We cannot remove the country_code condition due to the posted_before field.

Ups! I forgot the required attribute 🤓

jco-odoo replied:
@suv-bmya I don't understand.  In order for it to be readonly you should not be in draft state anyways.  I think you should only modify it when the state is in draft anyways, so I don't see why you need the extra condition.

suv-bmya replied:
Nothing! You're right...We don't need this condition.

jco-odoo commented on addons/l10n_cl/views/account_move_view.xml:
```diff
@@ -15,6 +15,31 @@
+        <field name="arch" type="xml">
+            <field name="l10n_latam_document_number" position="attributes">
+                <attribute name="attrs">
+                    {'invisible': [
```
invisible parameters seem correct, but the indentation of the first part can be improved.

suv-bmya replied:
it's weird 🤔

jco-odoo reviewed (changes requested):
@suv-bmya Should not take you a lot of to finish this.

suv-bmya commented:
> @suv-bmya Should not take you a lot of to finish this.

@jco-odoo . Done!

vbe-odoo commented:
@vbe-odoo do a FUN test and confirm

vbe-odoo commented:
@jco-odoo FUN test done, and LGTH.
Ready for technical review > merge.
Thanks.
cc' @suv-bmya @Danisan

jco-odoo commented:
@suv-bmya Could you put a standard format commit message: [FIX] l10n_latam_invoice_document: ....

suv-bmya commented:
> @suv-bmya Could you put a standard format commit message: [FIX] l10n_latam_invoice_document: ....

ups! Sure.

jco-odoo commented:
@robodoo r+

robodoo commented:
Linked pull request(s) odoo/enterprise#18289 not ready. Linked PRs are not staged until all of them are ready.
